modules/standard/Path.chpl

```diff
@@ -287,6 +288,85 @@ proc file.realPath(): string throws {
     }
   }
 
+/* Expands any environment variables in the path of the form `$<name>` or
+   `${<name>}` into their values.  If <name> does not exist, they are left
+   in place. Returns the path which includes these expansions.
+
+   :arg path: a string representation of a path, which may or may not include
+                   `$<name>` or `${<name>}`.
+   :type path: `string`
+
+   :return: `path`, having replaced all references to environment variables with
+                their values
+   :rtype: `string`
+*/
+  proc expandVars(path: string): string{
+    var path_p:string = path;
+    var varChars:string = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-_";
```

Comment:
I was all ready to comment that ``-`` isn't a legal character in an environment variable name.  I'm glad I checked the ``setenv`` man page before speaking up.  It turns out that any character is legal in an environment variable name, apart from ``=`` and the NULL character.  However, what to do about it is more nuanced.

bash at least doesn't accept anything but alphanumeric and ``_``.  But the ``setenv()`` libc call itself (and presumably ``getenv()``) do allow variables with other characters, as reported by ``/usr/bin/env``, and more immediately relevant, by python's ``os.path.expandvars()`` in some cases.

I'll spare you the experimentation by sharing the code I used to check.

First, a program to call ``setenv()``, and then exec another program to see what actually got set in the environment:

setenv.c:
```c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

int main(int argc, char **argv) {
  char *e = argv[1];
  char *v = argv[2];
  char *b;
  int rc;

  rc = setenv(e, v, 1);
  if (rc != 0) {
    fprintf(stderr, "setenv('%s', '%s') failed: %s\n",
	    e, v, strerror(errno));
    exit(1);
  }

  //rc = execl("/bin/bash", "bash", (char*)NULL);
  //rc = execl("/usr/bin/env", "env", (char*)NULL);
  b = malloc(strlen(e) + 4);
  if (b == NULL) {
    fprintf(stderr, "malloc failed:\n");
    exit(1);
  }
  sprintf(b, "${%s}", e);
   
  rc = execl("./expandvars.py", "expandvars.py", b, (char*)NULL);

  fprintf(stderr, "execl failed: %d: %s\n", rc, strerror(errno));
  free(b);
  return rc;
}

```

And the expandvars.py script to see what ``os.path.expandvars()`` does:

```python
#!/usr/bin/env python

import os.path
import argparse


def main():
    parser = argparse.ArgumentParser(description='call expandvars')
    parser.add_argument('vn', type=str, help="var name")
    args = parser.parse_args()

    startstr = "/foo/" + args.vn + "/bar"
    newstr = os.path.expandvars(startstr)
    print("got {newstr}".format(newstr=newstr))
    return 0


if __name__ == "__main__":
    main()
```

Observations:
* With expandvars.py execing /usr/bin/env, ``./setenv "SNARF SNARF" "GLEEP"`` invokes env which does include the line ``SNARF SNARF=GLEEP``.
* With expandvars.py execing bash, and running ``env`` within bash
** ``./setenv SNARF GLEEP`` does result in ``SNARF=GLEEP``, showing that bash startup isn't obliterating the new variables.
** But ``./setenv "SNARF SNARF" GLEEP`` doesn't get set, revealing (I think) that bash suppresses environment variables it doesn't like.
* With expandvars.py execing expandpath.py,
** ``./setenv "SNARF SNARF" GLEEP`` shows that ``os.path.expandvars()`` _does_ expand variables with spaces in them.  (Similarly, hyphens and dollar signs.)
*** But only in the ``${SNARF SNARF}`` or ``${SNARF-SNARF}`` syntax.  It doesn't expand ``$SNARF SNARF`` or ``$SNARF-SNARF``.  (``setenv.c`` code to invoke this not shown, but trivial.)


Which I think is pretty much what you've implemented, except for the ``-`` in ``varChars``.  So this is just a very long way of saying that I think there's one extra character in the commit.  :smile:

Reply:
I would try using a [regular expression](https://chapel-lang.org/docs/modules/standard/Regexp.html) here instead of listing all the characters out (note that using this support will require a `.skipif` file for your test when regex support is not built it, like [this file](https://github.com/chapel-lang/chapel/blob/master/test/release/examples/benchmarks/shootout/pidigits.skipif), but there aren't many situations where that support won't build)